File: csharp/incremental_updates/comments/comments.cs
// ============================================================================
//  comments -- C# port of examples\Vb6\incremental_updates\comments\comments.bas
//  Incremental updates: create a file with one square annotation in memory, then
//  repeatedly reply to the annotation (and to the reply) saving each step as an
//  incremental update, and finally write the result to disk.
// ============================================================================
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class Comments
{
    static TErrorProc _err = PDFError;

    static int PDFError(IntPtr data, int errCode, string errMessage, int errType)
    {
        Console.WriteLine(errMessage);
        return 0;
    }

    static bool CreateTestFile(IntPtr pdf, ref byte[] buf)
    {
        LumasPdf.pdfCreateNewPDFW(pdf, "");
        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);

        LumasPdf.pdfAppend(pdf);
        LumasPdf.pdfSquareAnnotW(pdf, 50.0, 50.0, 200.0, 100.0, 3.0, LumasPdfConsts.NO_COLOR, 255,
            TPDFColorSpace.csDeviceRGB, "Jim", "Test", "Just a test...");
        LumasPdf.pdfEndPage(pdf);

        if (!LumasPdf.pdfCloseFile(pdf)) return false;
        uint bufSize = 0;
        IntPtr p = LumasPdf.pdfGetBuffer(pdf, ref bufSize);
        if (p == IntPtr.Zero || bufSize == 0) return false;
        buf = new byte[bufSize];
        Marshal.Copy(p, buf, 0, (int)bufSize);
        // Release the original buffer; this also resets the PDF instance.
        LumasPdf.pdfFreePDF(pdf);
        return true;
    }

    static bool LoadTestFile(IntPtr pdf, byte[] buf)
    {
        LumasPdf.pdfCreateNewPDFW(pdf, "");
        LumasPdf.pdfSetImportFlags2(pdf, LumasPdfConsts.if2IncrementalUpd);
        GCHandle h = GCHandle.Alloc(buf, GCHandleType.Pinned);
        try
        {
            if (LumasPdf.pdfOpenImportBuffer(pdf, h.AddrOfPinnedObject(), (uint)buf.Length,
                    (int)LumasPdfConsts.ptOpen, "") < 0) return false;
            return LumasPdf.pdfImportPDFFile(pdf, 1, 1.0, 1.0) > 0;
        }
        finally { h.Free(); }
    }

    static bool SaveFile(IntPtr pdf, ref byte[] buf)
    {
        if (!LumasPdf.pdfCloseFile(pdf)) return false;
        uint bufSize = 0;
        IntPtr p = LumasPdf.pdfGetBuffer(pdf, ref bufSize);
        if (p == IntPtr.Zero || bufSize == 0) return false;
        buf = new byte[bufSize];
        Marshal.Copy(p, buf, 0, (int)bufSize);
        LumasPdf.pdfFreePDF(pdf);
        return true;
    }

    static void Main()
    {
        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);

        byte[] buf = null;
        if (!CreateTestFile(pdf, ref buf))
        {
            LumasPdf.pdfDeletePDF(pdf);
            return;
        }

        if (LoadTestFile(pdf, buf))
        {
            // The file contains only one annotation; its handle is zero (an array index).
            int reply = LumasPdf.pdfSetAnnotMigrationStateW(pdf, 0, TAnnotState.asCreateReply, "Harry");
            LumasPdf.pdfSetAnnotStringW(pdf, (uint)reply, (int)TAnnotString.asContent,
                "Hi Jim, your test annotation looks fine!");
            if (SaveFile(pdf, ref buf) && LoadTestFile(pdf, buf))
            {
                reply = LumasPdf.pdfSetAnnotMigrationStateW(pdf, (uint)reply, TAnnotState.asCreateReply, "Tommy");
                LumasPdf.pdfSetAnnotStringW(pdf, (uint)reply, (int)TAnnotString.asContent,
                    "Just a test whether I can reply to a reply...");
                if (SaveFile(pdf, ref buf) && LoadTestFile(pdf, buf))
                {
                    reply = LumasPdf.pdfSetAnnotMigrationStateW(pdf, (uint)reply, TAnnotState.asCreateReply, "Jim");
                    LumasPdf.pdfSetAnnotStringW(pdf, (uint)reply, (int)TAnnotString.asContent,
                        "Seems to work very well!");
                    if (LumasPdf.pdfHaveOpenDoc(pdf))
                    {
                        string filePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");
                        if (LumasPdf.pdfOpenOutputFileW(pdf, filePath))
                        {
                            if (LumasPdf.pdfCloseFile(pdf))
                                Console.WriteLine("PDF file \"" + filePath + "\" successfully created!");
                        }
                    }
                }
            }
        }

        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: cpp/incremental_updates/comments/comments.cpp
// comments -- C++ port of examples\Vb6\incremental_updates\comments\comments.bas
// Incremental updates: create a file with a square annotation, then reply to it
// (and to the reply) saving each step as an incremental update.
#include <lumaspdf.h>
#include <cstdio>
#include <cstring>
#include <string>
#include <vector>

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

// The Delphi original passes `string` (UnicodeString), which binds to the
// WideString overload of CreateNewPDF/SquareAnnot/SetAnnotMigrationState/
// SetAnnotString/OpenOutputFile, so the reference writes those annotation
// strings as UTF-16. (OpenImportBuffer has ONLY an AnsiString overload in
// dynapdf.pas, so it correctly stays on the A entry point below.)
static std::basic_string<LWCHAR> W(const std::string& s){
    return std::basic_string<LWCHAR>(s.begin(), s.end());
}

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    if(ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

static bool CreateTestFile(PPDF pdf, std::vector<unsigned char>& buf){
    pdfCreateNewPDFW(pdf, (LWCHAR*)LUMAS_TEXT(""));
    pdfSetPageCoords(pdf, pcTopDown);
    pdfAppend(pdf);
        pdfSquareAnnotW(pdf, 50.0, 50.0, 200.0, 100.0, 3.0, NO_COLOR, 255, csDeviceRGB,
                        (LWCHAR*)LUMAS_TEXT("Jim"), (LWCHAR*)LUMAS_TEXT("Test"), (LWCHAR*)LUMAS_TEXT("Just a test..."));
    pdfEndPage(pdf);
    if(pdfCloseFile(pdf) == 0) return false;
    UI32 sz = 0;
    char* p = pdfGetBuffer(pdf, &sz);
    if(!p || sz == 0) return false;
    buf.assign((unsigned char*)p, (unsigned char*)p + sz);
    pdfFreePDF(pdf); // releases the original buffer and resets the instance
    return true;
}

static bool LoadTestFile(PPDF pdf, std::vector<unsigned char>& buf){
    pdfCreateNewPDFW(pdf, (LWCHAR*)LUMAS_TEXT(""));
    pdfSetImportFlags2(pdf, if2IncrementalUpd);
    if(pdfOpenImportBuffer(pdf, buf.data(), (UI32)buf.size(), ptOpen, "") < 0) return false;
    return pdfImportPDFFile(pdf, 1, 1.0, 1.0) > 0;
}

static bool SaveFile(PPDF pdf, std::vector<unsigned char>& buf){
    if(pdfCloseFile(pdf) == 0) return false;
    UI32 sz = 0;
    char* p = pdfGetBuffer(pdf, &sz);
    if(!p || sz == 0) return false;
    buf.assign((unsigned char*)p, (unsigned char*)p + sz);
    pdfFreePDF(pdf);
    return true;
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, nullptr, PDFError);

    std::vector<unsigned char> buf;
    if(!CreateTestFile(pdf, buf)){ pdfDeletePDF(pdf); return 1; }

    if(LoadTestFile(pdf, buf)){
        SI32 reply = pdfSetAnnotMigrationStateW(pdf, 0, asCreateReply, (LWCHAR*)LUMAS_TEXT("Harry"));
        pdfSetAnnotStringW(pdf, reply, asContent, (LWCHAR*)LUMAS_TEXT("Hi Jim, your test annotation looks fine!"));
        if(SaveFile(pdf, buf) && LoadTestFile(pdf, buf)){
            reply = pdfSetAnnotMigrationStateW(pdf, reply, asCreateReply, (LWCHAR*)LUMAS_TEXT("Tommy"));
            pdfSetAnnotStringW(pdf, reply, asContent, (LWCHAR*)LUMAS_TEXT("Just a test whether I can reply to a reply..."));
            if(SaveFile(pdf, buf) && LoadTestFile(pdf, buf)){
                reply = pdfSetAnnotMigrationStateW(pdf, reply, asCreateReply, (LWCHAR*)LUMAS_TEXT("Jim"));
                pdfSetAnnotStringW(pdf, reply, asContent, (LWCHAR*)LUMAS_TEXT("Seems to work very well!"));
                if(pdfHaveOpenDoc(pdf) != 0){
                    std::string filePath = exeDir(argv[0]) + "/out.pdf";
                    if(pdfOpenOutputFileW(pdf, (LWCHAR*)W(filePath).c_str()) != 0){
                        if(pdfCloseFile(pdf) != 0)
                            printf("PDF file \"%s\" successfully created!\n", filePath.c_str());
                    }
                }
            }
        }
    }
    pdfDeletePDF(pdf);
    return 0;
}
